File: app/src/main/java/com/example/mydream/ProductDao.java
package com.example.mydream;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;

import java.util.List;

import static android.webkit.WebSettings.PluginState.ON;

@Dao
public interface ProductDao {
    @Insert
    public void Insert(Product product);

    @Delete
    public void Delete(Product product);

    @Query("SELECT * FROM PRODUCT WHERE itemId LIKE :itemId")
    public Product getProductDetail(int itemId);

    @Query("SELECT * FROM PRODUCT")
    public LiveData<List<Product>> getAllProduct();

    @Query(" SELECT * FROM PRODUCT JOIN PRODCUTFTS ON PRODUCT.itemName = ProdcutFts OR PRODUCT.itemSimilarName=PRODCUTFTS.itemSimilarName WHERE ProdcutFts MATCH :query ")
    public List<Product> autoSuggestion(String query);

//    @Query(" SELECT itemName FROM PRODUCT JOIN PRODCUTFTS ON PRODUCT.itemName = PRODCUTFTS.itemName  WHERE ProdcutFts MATCH '\"str*\"' LIMIT 10 ")
//    public List<String> autoSuggestion(String str);



}

File: app/src/main/java/com/example/mydream/ProductDao.kt
package com.example.mydream

import androidx.room.Dao
import com.example.mydream.Product
import androidx.room.Delete
import androidx.lifecycle.LiveData
import androidx.room.Insert
import androidx.room.Query
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

@Dao
interface ProductDao {
    @Insert
    suspend fun Insert(product: Product?)

    @Delete
    suspend fun Delete(product: Product?)

    @Query("SELECT * FROM PRODUCT WHERE itemId LIKE :itemId")
    suspend fun getProductDetail(itemId: Int): Product?

    @get:Query("SELECT * FROM PRODUCT")
    val allProduct: LiveData<List<Product?>?>?

    @Query(" SELECT * FROM PRODUCT JOIN PRODCUTFTS ON PRODUCT.itemName = ProdcutFts OR PRODUCT.itemSimilarName=PRODCUTFTS.itemSimilarName WHERE ProdcutFts MATCH :query ")
    fun autoSuggestion(query: String?): List<Product?>?

    //    @Query(" SELECT itemName FROM PRODUCT JOIN PRODCUTFTS ON PRODUCT.itemName = PRODCUTFTS.itemName  WHERE ProdcutFts MATCH '\"str*\"' LIMIT 10 ")
    @get:Query("SELECT MAX(itemId) FROM PRODUCT")
    val id: Int
    //    public List<String> autoSuggestion(String str);
}